soc/arm/nxp_imx/mimx8ml8_m7/soc.c

```diff
@@ -0,0 +1,383 @@
+	 *                                                                                  write/
+	 *                                                                                  read
+	 *                                                                                  allocate
+	 * 2            x           0           0            Device           not shareable
+	 *  Above are normal use settings, if your want to see more details or want to config
+	 * different inner/outter cache policy. Please refer to Table 4-55 /4-56 in arm cortex-M7
+	 * generic user guide <dui0646b_cortex_m7_dgug.pdf>
+	 * param SubRegionDisable  Sub-region disable field. 0=sub-region is enabled, 1=sub-region
+	 *                          is disabled.
+	 * param Size              Region size of the region to be configured. use
+	 *                          ARM_MPU_REGION_SIZE_xxx MACRO in mpu_armv7.h.
+	 */
+
+	/*
+	 * Region 0 [0x0000_0000 - 0x4000_0000]:
+	 * Memory with Device type, not executable, not shareable, non-cacheable.
+	 */
+	MPU->RBAR = ARM_MPU_RBAR(0, 0x00000000U);
+	MPU->RASR = ARM_MPU_RASR(1, ARM_MPU_AP_FULL, 0, 0, 0, 1, 0, ARM_MPU_REGION_SIZE_1GB);
+
+	/*
+	 * Region 1 TCML[0x0000_0000 - 0x0001_FFFF]:
+	 * Memory with Normal type, not shareable, non-cacheable
+	 */
+	MPU->RBAR = ARM_MPU_RBAR(1, 0x00000000U);
+	MPU->RASR = ARM_MPU_RASR(0, ARM_MPU_AP_FULL, 1, 0, 0, 0, 0, ARM_MPU_REGION_SIZE_128KB);
+
+	/*
+	 * Region 2 QSPI[0x0800_0000 - 0x0FFF_FFFF]:
+	 * Memory with Normal type, not shareable, cacheable
+	 */
+	MPU->RBAR = ARM_MPU_RBAR(2, 0x08000000U);
+	MPU->RASR = ARM_MPU_RASR(0, ARM_MPU_AP_FULL, 1, 0, 1, 1, 0, ARM_MPU_REGION_SIZE_128MB);
+
+	/*
+	 * Region 3 TCMU[0x2000_0000 - 0x2002_0000]:
+	 * Memory with Normal type, not shareable, non-cacheable
+	 */
+	MPU->RBAR = ARM_MPU_RBAR(3, 0x20000000U);
+	MPU->RASR = ARM_MPU_RASR(0, ARM_MPU_AP_FULL, 1, 0, 0, 0, 0, ARM_MPU_REGION_SIZE_128KB);
+
+	/*
+	 * Region 4 DDR[0x4000_0000 - 0x8000_0000]:
+	 * Memory with Normal type, not shareable, non-cacheable
+	 */
+	MPU->RBAR = ARM_MPU_RBAR(4, 0x40000000U);
+	MPU->RASR = ARM_MPU_RASR(0, ARM_MPU_AP_FULL, 1, 0, 0, 0, 0, ARM_MPU_REGION_SIZE_1GB);
+
+	/*
+	   Non-cacheable area is provided in DDR memory, the DDR region [0x80000000 ~ 0x81000000]
+	   (please see the imx8mp-evk-rpmsg.dts) totally 16MB is revserved for CM7 core. You can put
+	   global or static uninitialized variables in NonCacheable section(initialized variables in
+	   NonCacheable.init section) to make them uncacheable. Since the base address of MPU region
+	   should be multiples of region size, to make it simple, the MPU region 5 set the address
+	   space 0x80000000 ~ 0xBFFFFFFF to be non-cacheable. Then MPU region 6 set the text and
+	   data section to be cacheable if the program running on DDR. The cacheable area base
+	   address should be multiples of its size in linker file, they can be modified per your
```

Comment:
Zephyr has a `nocache` section that is linked to the memory region defined as `SRAM` by the linker files. Could you select DDR memory as the SRAM region, or does your non cacheable region need to be defined in a separate memory region from general RAM?

Reply:
@danieldegrasse - This part is a little tricky as it's an auxiliary core to the Cortex A cores in the i.MX8M Plus SoC. The documentation and SDK from NXP shows that the M7 core supports running from several different regions, and they provide separate example linker files for each.

From what I can tell, the M7 core supports using the `nocache` `SRAM` section provided by Zephyr (the DTCM) in addition to a second `nocache` `SRAM` section in the DDR region.

One important caveat is that the M7 core supports running entirely from the DDR, and in this case, the `FLASH` section must be marked as cacheable.

I'll play with the settings here and see if I can come up with something more modular to support the different modes.

Reply:
See the latest changes. I moved to using the Zephyr MPU driver and added a `mpu_regions.c` file to match the same configuration.

I also added a section to the linker script for the additional DDR region that is available.

Reply:
Just checked, those changes look good, thanks for adding them! See my comment below, to keep the commit history of Zephyr clean those changes need to be squashed into the prior commit.